### PR TITLE
feat(nm): implementation of IPv6 configuration

### PR DESCRIPTION
**Pull request by sfiorani**
This PR adds the configuration of IPv6 protocol in the NetworkManager bundle.

**New enums to map NM settings and Snapshots settings in Kura**

- Added `NMSettingIP6ConfigAddrGenMode` and `NMSettingIP6ConfigPrivacy` enums to map the `addr-gen-mode` and `ip6-privacy` NetworkManager's settings
- Added `KuraIp6ConfigurationMethod` and `KuraIp6Privacy` enums that maps Snapshots' values for previous NM's settings
- Added `KuraIp6ConfigurationMethod` to map NetworkManager's setting `method` for IPv6 protocol

**IPv6 protocol configuration**

In `org.eclipse.kura.nm.configuration.NMSettingsConverter`, the `buildIpv6Settings` method is now implemented: the IPv6 configuration can be set according to NetworkManager requirements.

**Related Issue:** This PR fixes/closes {issue number}

**Description of the solution adopted:** A more detailed description of the changes made to solve/close one or more issues. If the PR is simple and easy to understand this section can be skipped

**Screenshots:** If applicable, add screenshots to help explain your solution

**Manual Tests:** Optional description of the tests performed to check correct functioning of changes, useful for an efficient review

**Any side note on the changes made:** Description of any other change that has been made, which is not directly linked to the issue resolution [e.g. Code clean up/Sonar issue resolution]

### CODE REVIEW

**Comment by mattdibi** [comment]
@sfiorani You have a couple of test failures probably due to a missing `ip6.status` property. Don't worry about it for now since we decided to change how to handle the missing property as per our latest discussion.

Set `ip6.status` as `Optional` as we discussed and then fix the tests.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6AddressGenerationMode.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    EUI64,
+    STABLE_PRIVACY;
+
+    private static final List<KuraIp6AddressGenerationMode> ADDRESS_GEN_MODE = Arrays
+            .asList(KuraIp6AddressGenerationMode.EUI64, KuraIp6AddressGenerationMode.STABLE_PRIVACY);
+
+    public static Boolean isEnabled(KuraIp6AddressGenerationMode status) {
+        return ADDRESS_GEN_MODE.contains(status);
+    }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6AddressGenerationMode.java]: Why did you add all this `isEnabled` for the enums? Are they used somewhere? Aren't they somewhat redundant (especially in this case since **all** the enums mean that it `isEnabled`)?

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6Privacy.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        case "netIPv6PrivacyDisabled":
+            return KuraIp6Privacy.DISABLED;
+        case "netIPv6PrivacyEnabledPubAdd":
+            return KuraIp6Privacy.ENABLED_PUBLIC_ADD;
+        case "netIPv6PrivacyEnabledTempAdd":
+            return KuraIp6Privacy.ENABLED_TEMP_ADD;
+        default:
+            return KuraIp6Privacy.UNKNOWN;
+        }
+    }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6Privacy.java]: I wouldn't model an enum as "Unknown"... either this is a value we support or we throw an error.

```suggestion
    public static KuraIp6Privacy fromString(String status) {
        switch (status) {
        case "netIPv6PrivacyUnknown":
            return KuraIp6Privacy.UNKNOWN;
        case "netIPv6PrivacyDisabled":
            return KuraIp6Privacy.DISABLED;
        case "netIPv6PrivacyEnabledPubAdd":
            return KuraIp6Privacy.ENABLED_PUBLIC_ADD;
        case "netIPv6PrivacyEnabledTempAdd":
            return KuraIp6Privacy.ENABLED_TEMP_ADD;
        default:
            throw new IllegalArgumentException("bla bla");
        }
    }
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6ConfigurationMethod.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        switch (status) {
+        case "netIPv6MethodAuto":
+            return KuraIp6ConfigurationMethod.AUTO;
+        case "netIPv6MethodDhcp":
+            return KuraIp6ConfigurationMethod.DHCP;
+        case "netIPv6MethodManual":
+            return KuraIp6ConfigurationMethod.MANUAL;
+        default:
+            return KuraIp6ConfigurationMethod.UNKNOWN;
+
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6ConfigurationMethod.java]: I wouldn't model an enum as "Unknown"... either this is a value we support or we throw an error.

```suggestion
        default:
            throw new IllegalArgumentException("bla bla");

```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java [hunk 4]
```diff
@@ -147,10 +150,93 @@ public static Map<String, Variant<?>> buildIpv4Settings(NetworkProperties props,
     }
 
     public static Map<String, Variant<?>> buildIpv6Settings(NetworkProperties props, String deviceId) {
+
+        KuraIpStatus ip6Status = KuraIpStatus
+                .fromString(props.get(String.class, "net.interface.%s.config.ip6.status", deviceId));
+
+        if (ip6Status == KuraIpStatus.UNMANAGED) {
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java]:
```suggestion
        if (ip6Status == KuraIpStatus.UNMANAGED || ip6Status == KuraIpStatus.UNKNOWN) {
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java [hunk 5]
```diff
@@ -147,10 +150,93 @@ public static Map<String, Variant<?>> buildIpv4Settings(NetworkProperties props,
...
+            settings.put("addr-gen-mode", new Variant<>(
+                    KuraIp6AddressGenerationMode.toNMSettingIP6ConfigAddrGenMode(ipv6AddressGenerationMode).toInt32()));
+        }
+
+        Optional<String> privacy = props.getOpt(String.class, "net.interface.%s.config.ip6.privacy", deviceId);
+        if (privacy.isPresent()) {
+            KuraIp6Privacy ip6Privacy = KuraIp6Privacy.fromString(privacy.get());
+            settings.put("ip6-privacy",
+                    new Variant<>(KuraIp6Privacy.toNMSettingIP6ConfigPrivacy(ip6Privacy).toInt32()));
+        }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java]: Should these two parameters only be checked if `KuraIp6ConfigurationMethod` is set to `AUTO`? If `KuraIp6ConfigurationMethod` is set to `dhcp-only` or `manual` no address auto generation takes place and thus I don't think these parameters are used.

File: kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java [hunk 6]
```diff
@@ -75,6 +75,7 @@
 import org.freedesktop.dbus.connections.impl.DBusConnection;
 import org.freedesktop.dbus.exceptions.DBusException;
 import org.freedesktop.dbus.exceptions.DBusExecutionException;
+import org.freedesktop.dbus.interfaces.ObjectManager;
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java]: I don't think this is used... do a double check please

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java [hunk 7]
```diff
@@ -147,10 +150,99 @@ public static Map<String, Variant<?>> buildIpv4Settings(NetworkProperties props,
...
+                settings.put("dns", new Variant<>(convertIp6(dnsServers.get()), "aay"));
+                settings.put("ignore-auto-dns", new Variant<>(true));
+            }
+
+            Optional<Integer> wanPriority = props.getOpt(Integer.class, "net.interface.%s.config.ip6.wan.priority",
+                    deviceId);
+            if (wanPriority.isPresent()) {
+                Long supportedByNM = wanPriority.get().longValue();
+                settings.put("route-metric", new Variant<>(supportedByNM));
+            }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java]: Use `Optional.ifPresent` instead of branching everytime. This way Sonar shouldn't complain about complexity. Example:

```suggestion
            wanPriority.ifPresent(value -> 
                settings.put("route-metric", new Variant<>(value.longValue()))
            )
```

**Comment by MMaiero** [comment]
Please check and solve the two issues reported by Sonar: https://sonarcloud.io/project/issues?resolved=false&severities=CRITICAL&sinceLeakPeriod=true&branch=PR-4786&id=org.eclipse.kura%3Akura

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java [hunk 8]
```diff
@@ -147,10 +150,99 @@ public static Map<String, Variant<?>> buildIpv4Settings(NetworkProperties props,
...
+                settings.put("addr-gen-mode", new Variant<>(KuraIp6AddressGenerationMode
+                        .toNMSettingIP6ConfigAddrGenMode(ipv6AddressGenerationMode).toInt32()));
+            }
+
+            Optional<String> privacy = props.getOpt(String.class, "net.interface.%s.config.ip6.privacy", deviceId);
+            if (privacy.isPresent()) {
+                KuraIp6Privacy ip6Privacy = KuraIp6Privacy.fromString(privacy.get());
+                settings.put("ip6-privacy",
+                        new Variant<>(KuraIp6Privacy.toNMSettingIP6ConfigPrivacy(ip6Privacy).toInt32()));
+            }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java]: Same here:

```suggestion
            privacy.ifPresent(value -> {
                KuraIp6Privacy ip6Privacy = KuraIp6Privacy.fromString(value);
                settings.put("ip6-privacy",
                        new Variant<>(KuraIp6Privacy.toNMSettingIP6ConfigPrivacy(ip6Privacy).toInt32()));
            });
```

File: kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/configuration/NMSettingsConverterTest.java [hunk 9]
```diff
@@ -1382,4 +1709,27 @@ public Object buildAddressDataWith(String ipAddr, UInt32 prefix) {
...
+
+        List<Byte> dnsAddress = new ArrayList<>();
+        for (int i = 0; i < dnsByteArray.length; i++) {
+            dnsAddress.add(dnsByteArray[i]);
+        }
+
+        dnsAddresses.add(dnsAddress);
+
+        return dnsAddresses;
+    }
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/configuration/NMSettingsConverterTest.java]: This is wrong on a fundamental level.

You cannot test your code with the code you're testing. You're replicating the logic you've written in the production code in your test. Doing so you're defeating the purpose of the test. See [our docs](https://eurotech.atlassian.net/wiki/spaces/ECE/pages/2271510531/Mattia+and+Greg+s+Unit+Testing+Tips#Guideline-2:-Prefer-static-data-in-expected-values)

> If we replicate the code logic in our tests we would need unit tests for our unit tests. The unit tests should be our single source of truth about the expected output given an input.

Substitute this method call with a constant wherever you used this method and get rid of this method.

_Note_: Yes, I noticed that there's a complementary method `buildAddressDataWith` that performs the same. That method is wrong too. We should weed that out. Ignore it for now, we'll take care of that in a future PR.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java [hunk 10]
```diff
@@ -123,14 +127,14 @@ public static Map<String, Variant<?>> buildIpv4Settings(NetworkProperties props,
         }
 
         if (ip4Status.equals(KuraIpStatus.ENABLEDLAN)) {
-            settings.put("ignore-auto-dns", new Variant<>(true));
+            settings.put(NM_SETTINGS_IPV6_IGNORE_AUTO_DNS, new Variant<>(true));
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java]: These are ipv4 settings. Create a dedicated different constant.

```suggestion
            settings.put(NM_SETTINGS_IPV4_IGNORE_AUTO_DNS, new Variant<>(true));
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java [hunk 11]
```diff
@@ -123,14 +127,14 @@ public static Map<String, Variant<?>> buildIpv4Settings(NetworkProperties props,
...
             settings.put("ignore-auto-routes", new Variant<>(true));
         } else if (ip4Status.equals(KuraIpStatus.ENABLEDWAN)) {
             Optional<List<String>> dnsServers = props.getOptStringList("net.interface.%s.config.ip4.dnsServers",
                     deviceId);
             if (dnsServers.isPresent()) {
                 settings.put("dns", new Variant<>(convertIp4(dnsServers.get()), "au"));
-                settings.put("ignore-auto-dns", new Variant<>(true));
+                settings.put(NM_SETTINGS_IPV6_IGNORE_AUTO_DNS, new Variant<>(true));
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/NMSettingsConverter.java]: These are ipv4 settings. Create a dedicated different constant.

```suggestion
                settings.put(NM_SETTINGS_IPV6_IGNORE_AUTO_DNS, new Variant<>(true));
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6AddressGenerationMode.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
+package org.eclipse.kura.nm;
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6AddressGenerationMode.java:13]: Missing copyright header

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6Privacy.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
+package org.eclipse.kura.nm;
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6Privacy.java:13]: Missing copyright header

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/enums/NMSettingIP6ConfigAddrGenMode.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
+package org.eclipse.kura.nm.enums;
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/enums/NMSettingIP6ConfigAddrGenMode.java:13]: Missing copyright header

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/enums/NMSettingIP6ConfigPrivacy.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
+package org.eclipse.kura.nm.enums;
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/enums/NMSettingIP6ConfigPrivacy.java:13]: Missing copyright header

File: kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/configuration/NMSettingsConverterTest.java [hunk 16]
```diff
@@ -1367,7 +1694,7 @@ public void thenNoExceptionsHaveBeenThrown() {
     }
 
     /*
-     * Helper Methods
+     * Helper Enables
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/configuration/NMSettingsConverterTest.java]: Uh?

File: kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/configuration/NMSettingsConverterTest.java [hunk 17]
```diff
@@ -51,6 +51,9 @@ public class NMSettingsConverterTest {
     Boolean hasAnIllegalArgumentExceptionThrown = false;
     Boolean hasAGenericExecptionBeenThrown = false;
 
+    private static List<List<Byte>> expectedDnsList = Arrays
+            .asList(Arrays.asList(new Byte[] { 32, 1, 72, 96, 72, 96, 0, 0, 0, 0, 0, 0, 0, 0, -120, 68 }));
+
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/configuration/NMSettingsConverterTest.java]: It is a constant therefore you'll need the `final` keyword. Furthermore you'll need to write it as all capital.

Finally I prefer to pass the expectation method a list of constants for readability. This means that this becomes:

 ```suggestion
    private static final List<Byte> IP6_BYTE_ARRAY_ADDRESS = Arrays.asList(new Byte[] { 32, 1, 72, 96, 72, 96, 0, 0, 0, 0, 0, 0, 0, 0, -120, 68 });

```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6Privacy.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        case "netIPv6PrivacyDisabled":
+            return KuraIp6Privacy.DISABLED;
+        case "netIPv6PrivacyEnabledPubAdd":
+            return KuraIp6Privacy.ENABLED_PUBLIC_ADD;
+        case "netIPv6PrivacyEnabledTempAdd":
+            return KuraIp6Privacy.ENABLED_TEMP_ADD;
+        default:
+            throw new IllegalArgumentException(String.format("Unsupported IPv6 privacy value: \"%s\"", status));
+        }
+    }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/KuraIp6Privacy.java]: Sorry to be back on this but I just noticed:
1. I don't think you need to use have an `unknown` case. Either is disabled or one of the enabled.
2. I would prefer more descriptive strings see below:

```suggestion
    DISABLED,
    ENABLED_PUBLIC_ADD,
    ENABLED_TEMP_ADD;

    public static KuraIp6Privacy fromString(String status) {
        switch (status) {
        case "netIPv6PrivacyDisabled":
            return KuraIp6Privacy.DISABLED;
        case "netIPv6PrivacyEnabledPublicAddress":
            return KuraIp6Privacy.ENABLED_PUBLIC_ADDR;
        case "netIPv6PrivacyEnabledTemporaryAddress":
            return KuraIp6Privacy.ENABLED_TEMP_ADDR;
        default:
            throw new IllegalArgumentException(String.format("Unsupported IPv6 privacy value: \"%s\"", status));
        }
    }
```

**Comment by mattdibi** [comment]
@sfiorani the only missing thing at this point are the unit tests. The production code is 👌

**Review by mattdibi — APPROVED** [review]

**Comment by MMaiero** [comment]
Opened IP review: https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/9819

**Comment by MMaiero** [comment]
IP check successful
